src/js/globals.d.ts

```diff
@@ -208,3 +216,23 @@ declare module "worker-loader?inline=true&fallback=false!*" {
 
     export default WebpackWorker;
 }
+
+// JSX
+declare namespace JSX {
+    // The return type of our JSX Factory: this could be anything
+    type Element = HTMLElement;
+
+    // IntrinsicElementMap grabs all the standard HTML tags in the TS DOM lib.
+    interface IntrinsicElements extends IntrinsicElementMap {}
+
+    // The following are custom types, not part of TS's known JSX namespace:
+    type IntrinsicElementMap = {
+        [K in keyof HTMLElementTagNameMap]: {
+            [k: string]: any;
+        };
+    };
```

Comment:
I wonder if typing could be improved, but it would probably be hacky, e.g.
> infer valid html attributes by looking at HTMLXyzElement interface, exclude all functions, and keep boolean and number attributes and treat everything else (including objects) as string attributes
```ts
type IntrinsicElementMap = {
    [K in keyof HTMLElementTagNameMap]: {
        [L in keyof Pick<
            HTMLElementTagNameMap[K],
            {
                [L in keyof HTMLElementTagNameMap[K]]: HTMLElementTagNameMap[K][L] extends Function
                    ? never
                    : L;
            }[keyof HTMLElementTagNameMap[K]]
        >]?: HTMLElementTagNameMap[K][L] extends boolean | number ? HTMLElementTagNameMap[K][L] : string;
    };
};
```

Reply:
I wouldn't consider this too hacky, but it could definitely be better (Especially the part about removing functions). I would propose something like this:

```ts
type IntrinsicElementMap = {
        [K in keyof HTMLElementTagNameMap]: {
            [L in keyof HTMLElementTagNameMap[K] as HTMLElementTagNameMap[K][L] extends Function ? never : L]?: 
                HTMLElementTagNameMap[K][L] extends boolean | number
                    ? HTMLElementTagNameMap[K][L]
                    : string;
        };
    };
```
Or perhaps even using a util type to make it more readable:
```ts
type OmitWhere<T, V> = {
    [K in keyof T as T[K] extends V ? never: K ]: T[K]
}

type IntrinsicElementMap2 = {
    [K in keyof HTMLElementTagNameMap]: {
        [L in keyof OmitWhere<HTMLElementTagNameMap[K], Function>]?: 
            HTMLElementTagNameMap[K][L] extends boolean | number ? HTMLElementTagNameMap[K][L] : string;
    }
};
```

Your suggestion also forbids any custom properties (`data-xyz`) for example, so do you oppose appending `& Record<string, string>` to each element? This does not remove typing for the keys that already exist but allows custom keys:
```ts
type IntrinsicElementMap3 = {
    [K in keyof HTMLElementTagNameMap]: {
        // ...
    } & Record<string, string>
};
```

Reply:
oh, good to know `in...as` syntax exists, that's a lot nicer! And yeah the helper type makes sense, though maybe it should be called `OmitByValue` for clarity?

also, as for custom properties, funny thing, according to <https://www.typescriptlang.org/docs/handbook/jsx.html#attribute-type-checking>:
> Note: If an attribute name is not a valid JS identifier (like a data-* attribute), it is not considered to be an error if it is not found in the element attributes type.

(Note that, even for hyphenated names, an exact property name match will error if the value's type isn't assignable.)

Though this makes me realize, `JSXCreateElement` assumes hyphenated names, while `HTMLElementTagNameMap` of course provides the JS camel-cased versions. ~~Helper type that maps camel case to kebab case???~~

Reply:
Here is a possible implementation of said type:
```ts
type uppercase = 'A' | 'B' | 'C' | 'D' | 'E' | 'F' | 'G' | 'H' | 'I' | 'J' | 'K' | 'L' | 'M' | 'N' | 'O' | 'P' | 'Q' | 'R' | 'S' | 'T' | 'U' | 'V' | 'W' | 'X' | 'Y' | 'Z';
type CamelCaseToKebabCase<str> = str extends `${infer word}${uppercase}${string}`
    ? word extends `${string}${uppercase}${string}`
        ? never
        : str extends `${word}${infer char}${infer rest}`
            ? `${word}-${Lowercase<char>}${CamelCaseToKebabCase<rest>}`
            : never
    : str;
type res = CamelCaseToKebabCase<"whatIsThisPlace">; // "what-is-this-place"
```
Is this worth it 🤔

Reply:
Actually, it looks like most multiword HTML properties (except `data-` and `aria-` and for some reason `accept-charset`) are straight lowercase (https://www.w3schools.com/tags/ref_attributes.asp), so `Lowercase<prop>` should be good in most cases no?

Reply:
yeah, I noted that from <https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes>. They aren't deprecated, unfortunately!

The hyphens are still necessary, though.

Reply:
On a separate note, maybe just directly assigning object properties is the way to go? It of course matches `HTMLElementTagNameMap`, so it's entirely consistent with JS but inconsistent with HTML.

Reply:
That would fail with `data-*` props though no?

a.dataFooBar = "a";
a.dataset.fooBar; // undefined

It also wouldn't be persistent if you got the element in another way later (`querySelector`)

Reply:
those would be assigned separately (I think it suffices to check if it starts with `data-`)

What wouldn't be persistent?

Reply:
I was thinking custom attributes, but it seems that (at least chrome) does actually keep them somewhere

Reply:
chose to be consistent with HTML, since the JSX is supposed to replace HTML anyway (and that also means using `setAttribute` and only allowing strings, numbers, and booleans)

inferring typings from JS properties is too hacky in multiple ways (especially dealing with hyphens vs. camel case), and if a better solution emerges in the future, it can be used